[숫자 야구 게임] 박상혁 미션 제출합니다.

sangcci opened the pull request:
## 개요
기존 코드에서 MVC 패턴 적용 리팩토링 진행

기존의 객체 설계는 책임이 너무 중구난방하게 있어 이를 일목요연하게 책임을 다시 부여 하였고 그 과정에서 MVC 패턴을 도입해서 효율적인 객체 설계를 완성했습니다.

pull request #2에 해당합니다.

## PR 유형
- [x] 패키지 재 구조화, 코드 리팩토링
- [x] README.md 기능 목록 문서 수정
- [x] 테스트 추가, 테스트 리팩토링

## 기능 목록
- [x] 게임 시작하기
- [x] 숫자 생성하기
- [x] 사용자로부터 Console 입력 받기
- [x] 숫자 검증하기
- [x] 타입 변환하기
- [x] 스트라이크, 볼, 낫싱 판단하기
- [x] Console 출력 하기
- [x] 게임 반복하기
- [x] 게임 종료하기
- [x] 다시 시작 여부 묻기 및 게임 다시 시작하기

sangcci committed:
docs: 객체 설계 및 기능 목록 추가

sangcci committed:
feat: "컴퓨터가 생성한 숫자가_3자리 서로 다른 임의의 숫자이면_성공한다" 테스트 추가

sangcci committed:
refactor($Computer): generateRandomNumber() 메서드 로직 수정

- 각 자리수 비교할 때 조건에 맞지 않으면 숫자 통째로 재생성

sangcci committed:
refactor($Computer): generateRandomNumber() 내부 로직 메서드 분리

sangcci committed:
feat($NumberBaseballTest, Game): "사용자로부터 받은 숫자가_3자리 서로다른 숫자가 아닐경우_예외를 발생한다." 테스트 추가

sangcci committed:
refactor($NumberBaseballTest): Game, Computer 객체를 필드로 변경

sangcci committed:
refactor($NumberBaseballTest): assertThatThrownBy() import static 처리

sangcci committed:
feat($Game, NumberBaseballTest): "사용자 숫자를 넣고_컴퓨터 숫자와 모두 같다면_3 스트라이크를 반환한다" 테스트 추가

sangcci committed:
refactor($NumberBaseballTest): 직전 테스트 이름 변경

sangcci committed:
fix($NumberBaseballTest, Game): 출력 문 띄어쓰기 수정

- "3 스트라이크" -> "3스트라이크"

sangcci committed:
fix($NumberBaseballTest, Game): 각 자리 같은 숫자 예시 변경

- "111"은 존재할 수 없다 -> "123"

sangcci committed:
feat($NumberBaseballTest, Game): "사용자 숫자를 넣고_컴퓨터 숫자와 1자리 자리까지 같고 2자리 똑같은 수가 있다면_2볼 1스트라이크를 반환한다" 테스트 추가

sangcci committed:
test($NumberBaseballTest): 테스트 범위 재정의로 인한 테스트 이름 수정

- 숫자를 비교하는 단위 테스트가 너무 많아져서 축소하였다
- 테스트 네이밍 수정
- 테스트 케이스 추가

sangcci committed:
refactor($Game): 메서드 추출

- compareNumber()에서 출력 만들기 부분 메서드 추출

sangcci committed:
refactor($Game): 메서드 추출

- compareNumber()에서 볼, 스트라이크 계산 로직 메서드로 추출

sangcci committed:
feat($NumberBaseballTest, Game): "사용자 숫자를 넣고_3스트라이크이면_게임을 종료한다" 테스트 추가

sangcci committed:
feat($NumberBaseballTest, Game): "게임을 종료한 후 다시시작 여부를 물을 때_1을 반환하면_게임을 다시시작한다" 테스트 추가

sangcci committed:
refactor($Computer, Game): 구현 객체 main으로 이동

sangcci committed:
refactor($Input, Output): 입력 및 출력 객체 생성

sangcci committed:
refactor($Game): start(), restart() 수정

sangcci committed:
feat($Application): main 로직 작성

sangcci committed:
fix($Computer): numberToString() 로직 수정

- 테스트 코드 로직은 pickNumberInRange() 메서드를 각 자리마다 총 3번 사용해서 구현한다. 이에 3번 사용하도록 수정

sangcci committed:
fix($Game): computerNumber 생성 로직 위치 수정

- compareNumber() 메서드에 위치하다보니, 한 번 게임할 때 마다 계속 computerNumber가 갱신되는 문제 발생
- 이를 start() 로직으로 옮김으로써 해결

sangcci committed:
fix($Input): 입력 라이브러리 수정

- 우아한 테크코스 전용 라이브러리인 Console로 수정

sangcci committed:
feat($Game): 3자리의 숫자가 아닐 경우 검증 로직 추가

- validateUserNumber()

sangcci committed:
fix($NumberBaseballTest): 로직 변화에 따른 테스트 주석 처리

sangcci committed:
Merge pull request #1 from baksakcci/develop

숫자 야구 게임 구현

sangcci committed:
docs: README.md 기능 명세서 수정

sangcci committed:
feat: 게임 시작하기 구현

sangcci committed:
feat: 컴퓨터 숫자 생성하기, console 출력하기 구현

sangcci committed:
refactor: 정적 팩토리 메서드 추가

- generateRandomNumber()를 정적 팩토리 메서드로 변경

sangcci committed:
test: 숫자 검증 테스트 케이스 추가

sangcci committed:
feat: 숫자 검증하기 구현

sangcci committed:
refactor: 메서드 이름 수정

- print -> displayString

sangcci committed:
refactor($Computer): generateRandomNumber() 내부 로직 수정

- computerNumber 타입을 List<Integer>로 변경

sangcci committed:
refactor($OutputView): List<Integer> 타입 출력 메서드 추가

sangcci committed:
refactor($NBGameController): 타입 변화에 따른 출력 메서드 변경

sangcci committed:
style($NBGameController): 안쓰는 import문 삭제

sangcci committed:
feat: 사용자로부터 Console 입력 받기 구현

sangcci committed:
refactor: isNumber 접근제어자 변경

- public -> private

sangcci committed:
feat: 숫자 검증 로직 추가

sangcci committed:
fix: 검증 논리 오류 수정

sangcci committed:
docs: README.md 수정

- 숫자 검증하기 완료

sangcci committed:
docs: 기능 목록에 타입 변환하기 추가

sangcci committed:
test: 타입 변환하기 테스트 케이스 추가

sangcci committed:
feat: 타입 변환하기 구현 완료

sangcci committed:
test: 스트라이크, 볼, 낫싱 판단하기 테스트 케이스 추가

sangcci committed:
feat: 스트라이크, 볼, 낫싱 판단하기 구현 완료

sangcci committed:
feat: 타입 변경 및 스트라이크 판단 로직 Controller 추가

- 통합 테스트(프로그램 테스트) 완료

sangcci committed:
feat: 결과 출력 추가

- displayResult() 추가

sangcci committed:
fix: 이전 클래스 삭제

sangcci committed:
feat: 게임 반복하기 구현

- 3스트라이크 나올 때 까지 게임 반복

sangcci committed:
docs: README 게임 반복하기 구현 반영

sangcci committed:
refactor: 객체 생성 코드 생성자로 옮김, run() static 메서드 구현

sangcci committed:
refactor: 게임 시작 메서드 합침, static 메서드 제거 및 메서드 이름 변경

sangcci committed:
test: 다시 시작 숫자 검증 테스트 케이스 추가

sangcci committed:
feat: 다시 시작 숫자 검증 추가

sangcci committed:
feat: 다시 시작 여부 묻기 및 게임 다시 시작하기 구현

sangcci committed:
refactor: displayList() 메서드 제거

sangcci committed:
refactor: displayList() 메서드 제거 및 주석 제거

sangcci committed:
refactor: view 메서드 추가

- view의 역할을 확대.

sangcci committed:
Merge pull request #2 from baksakcci/develop

숫자 야구 게임 MVC 패턴 적용

ChoiWonYu commented on src/main/java/baseball/controller/NBGameController.java:
```diff
@@ -0,0 +1,69 @@
+    private static final int RESTART_GAME = 1;
+    private static final int END_GAME = 2;
+
+    public NBGameController() {
```
정적 팩토리 메서드를 활용하는 건 어떨까요??

sangcci replied:
이름을 마음대로 지을 수 있다는 점에서 정적 팩토리 메서드를 활용할 수 있겠네요! 의견 감사합니다!

ChoiWonYu commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,28 @@
+
+    public static Computer generateRandomNumber() {
+        List<Integer> computer = new ArrayList<>();
+        while (computer.size() < 3) {
```
3과 같은 매직 넘버는 상수화 하는 게 어떨까요? 다른 숫자로 변경됐을 때 수정해야할 부분이 많을 것 같아요!

sangcci replied:
이 부분을 캐치 못했네요. 의견 감사합니다 :)

ChoiWonYu commented on src/main/java/baseball/model/validator/NumberValidation.java:
```diff
@@ -0,0 +1,27 @@
+        return input.length() != 3;
+    }
+
+    private boolean isDuplicateNumber(String userNumber) {
```
Set을 활용한 중복 숫자 검사 방법은 어떨까요??

sangcci replied:
미쳐 생각하지 못한 부분이였습니다! 한 수 배웁니다..!

ChoiWonYu commented on src/main/java/baseball/view/OutputView.java:
```diff
@@ -0,0 +1,37 @@
+        System.out.println(endMessage);
+    }
+
+    public void displayResult(int strike, int ball) {
```
dto에서 strike와 ball 개수를 이용해 출력문을 View에 넘겨주는 방법은 어떨까요?

sangcci replied:
말씀하신 대로 View에서 담당하는 책임은 아닌 것 같네요..! 미리 Model에서 출력문을 만든 후에 View에 넘겨줘야 겠습니다!

ChoiWonYu reviewed (commented):
코드 잘봤습니다~ 1주차 고생하셨습니다:)

reddevilmidzy commented on src/main/java/baseball/Application.java:
```diff
@@ -1,7 +1,11 @@
+
     public static void main(String[] args) {
-        // TODO: 프로그램 구현
+        NBGameController gameController = new NBGameController();
```
혹시 NB가 어떤 약자일까요??

reddevilmidzy replied:
아 NumberBaseball 인가요?
일반적이지 않은 약어는 한번에 이해하기 어려울 수도 있으니 전체이름 사용하는 것도 나쁘진 않을 거 같아요..!

sangcci replied:
이름이 너무 길어질 것 같아서 축약해서 사용했는데, 다시 생각해보니 리뷰어 입장에서 한 번에 이해하기 힘들 것 같네요..!
이름 짓는게 제일 힘든 것 같습니다..ㅠ 의견 감사합니다!

reddevilmidzy commented on src/main/java/baseball/model/BaseballCount.java:
```diff
@@ -0,0 +1,46 @@
+
+    public BaseballCount() {
+        hitterCount = new HashMap<>();
+        hitterCount.put("ball", 0);
+        hitterCount.put("strike", 0);
```
키값을 상수로 관리하는건 어떨까요?
언제 어디서 strike나 ball에 오타가 발생할 수도...!

sangcci replied:
String이라는 것을 인지를 못했네요..! 상수로 관리하면 실수의 여지를 없앨 수 있을 것 같습니다! 한 수 배웁니다.

reddevilmidzy commented on src/main/java/baseball/model/validator/RestartNumberValidation.java:
```diff
@@ -0,0 +1,17 @@
+
+public class RestartNumberValidation {
+
+    private static final Integer RESTART_GAME = 1;
+    private static final Integer END_GAME = 2;
```
여기 1과 2는 컨트롤러에서 사용하는 1과 2랑 같은 숫자이기에 한곳에서 관리하는 것도 좋을 거 같아요

sangcci replied:
상수 값이 중복이 발생했네요.. 의견 감사합니다

reddevilmidzy commented on src/main/java/baseball/view/OutputView.java:
```diff
@@ -0,0 +1,37 @@
+        }
+    }
+
+    private boolean isNothing(int strike, int ball) {
+        return strike == 0 && ball == 0;
+    }
+
+    private boolean isZero(int count) {
+        return count == 0;
+    }
```
이런 로직이 View에는 안어울리는 거 같아요...!

sangcci replied:
말씀하신 대로 View에서 Model의 책임이 넘어갔네요..! 이를 Model에서 구현하고 출력문만 넘겨주게 수정해야 겠습니다.. 의견 감사합니다!

reddevilmidzy commented on src/main/java/baseball/model/BaseballCount.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    private int calculateStrikeCount(List<Integer> userNumber, List<Integer> computerNumber) {
+        int strike = 0;
+        for (int i = 0; i < 3; i++) {
+            if (userNumber.get(i).equals(computerNumber.get(i))) {
+                strike++;
+            }
+        }
+        return strike;
```
스트림을 사용해서 위 로직을 처리해도 좋을 거 같아요!

sangcci replied:
맞네요! 스트림을 사용하면 코드가 줄어들어 보기 편할 것 같습니다! 감사합니다 :)

reddevilmidzy commented on src/main/java/baseball/model/BaseballCount.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public Map<String, Integer> pitch(List<Integer> user, List<Integer> computer) {
+        int ballCount = calculateBallCount(user, computer);
+        int strikeCount = calculateStrikeCount(user, computer);
+        ballCount -= strikeCount;
```
딱 봤을 때 ballCount 에서 왜 strikeCount 를 빼는지 이해가 안될 수도 있을거 같아요

저도 볼 갯수 구할 때 같은 위치에 있나 확인하는 로직을 넣는게 아니라 스트라이크 갯수를 빼는 식으로도 할 수 있겠다고 생각했었는데, 위에서 말했듯이 처음 보는 사람이 왜 ballCount에서 strikeCount 를 뺀 값이 ballCount인지 잘 이해가 안될거 같아서 ball 구하는 로직을 따로 만들었습니다

sangcci replied:
맞는 것 같습니다! 저런 로직은 `calculateBallCount`에 넣는 것이 리뷰어 입장에서 이해하기 쉬울 것 같습니다. 감사합니다!

reddevilmidzy reviewed (commented):
수고 많으셨습니다!

JeonSungGwon commented on src/main/java/baseball/controller/NBGameController.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    private boolean is3Strike(int strike) {
+        return strike != 3;
```
숫자 3도 상수로 지정해주면 더 좋을 거 같아요!

sangcci replied:
앗 여기를 놓쳤었네요 감사합니다 :)

JeonSungGwon reviewed (commented):
코드에 열정이 잘 보이는 거 같아요! 기능 분리도 잘 하셨고 무엇보다 테스트코드를 디테일하게 잘 구현하셨네요. 1주차 고생 많으셨습니다!

seokmyungham commented:
세심하게 신경 쓴 부분들이 많이 보이는 코드들이네요
실력이 뛰어나신 것 같아요. 잘봤습니다

seokmyungham commented on src/main/java/baseball/model/BaseballCount.java:
```diff
@@ -0,0 +1,46 @@
+        int ball = 0;
+        for (int i = 0; i < 3; i++) {
+            if (userNumber.contains(computerNumber.get(i))) {
+                ball++;
```
모델 객체 내에서만 사용되는 메서드들은 접근 지정자를 private로 설정하신 것 인상깊네요. 한 수 배워갑니다

seokmyungham commented on src/main/java/baseball/model/validator/NumberValidation.java:
```diff
@@ -0,0 +1,27 @@
+        } else if (isDuplicateNumber(userNumber)) {
+            throw new IllegalArgumentException("각 자리에 중복된 숫자가 있다");
+        }
+    }
```
지금도 가독성있게 코드를 잘 작성하셨지만
else 예약어 대신에 검증 목적을 의미있는 네이밍의 메서드로 각각 분리하는 방법은 어떨까요?
향후 유지 보수 측면에서 도움이 될 것 같다 생각됩니다 ㅎㅎ

seokmyungham commented on src/main/java/baseball/model/BaseballCount.java:
```diff
@@ -0,0 +1,46 @@
+
+    private int calculateBallCount(List<Integer> userNumber, List<Integer> computerNumber) {
+        int ball = 0;
+        for (int i = 0; i < 3; i++) {
```
모델 객체 내에서만 사용되는 메서드들은 접근 지정자를 private로 설정하신 것 인상깊네요. 한 수 배워갑니다

sangcci committed:
refactor: naming 수정

- NBGameController -> GameController

sangcci committed:
refactor: BaseballNumber VO 객체 추가로 인한 숫자 검증 로직 통합

sangcci committed:
fix: 추상 팩토리 메서드 추가

sangcci committed:
refactor: 입력값과 Random값 생성을 염두하여 String, List<Integer> 두가지 생성자 생성

sangcci committed:
refactor: Computer 대신 숫자를 생성하는 Pitcher 역할 추가

sangcci committed:
fix: 0이 있는지 검증 내부 로직 수정

sangcci committed:
refactor: 스트라이크, 볼을 판단하는 BaseballCount를 새로 Umpire라는 객체로 개편

sangcci committed:
refactor: controller에 Umpire 적용

sangcci committed:
refactor: 사용하지 않는 클래스 제거

sangcci committed:
refactor: 게임 재시작 역할을 맡은 객체 생성

sangcci committed:
refactor: RestartNumber 적용 및 while 에서의 제어 플래그 제거

sangcci committed:
refactor: RestartNumber 제거 및 RestartState 1개로 운용

sangcci committed:
refactor: 함수형 인터페이스 삭제

sangcci committed:
refactor: determineStrikeAndBall 메서드의 출력 부분을 getter로 분리

sangcci committed:
refactor: getter로 list를 출력하는 대신, strike, ball 각각 출력하도록 변경

sangcci committed:
refactor: ResultType 으로 출력 구문 구분

sangcci committed:
test: RestartTest 추가

sangcci committed:
test: 3스트라이크 판단 utility 테스트 추가

- Assertions static import화
- 좀 더 보기 쉽게 displayName 수정
- @ParameterizedTest의 name 속성 이용

sangcci committed:
refactor: Assertions static import화

sangcci committed:
style: 안쓰는 import문 제거

sangcci committed:
feat($BaseballNumber): 숫자 검증 추가

sangcci committed:
test: BaseballNumber 테스트 추가

sangcci committed:
feat($GameController): 의존성 객체에 final 추가

sangcci committed:
fix($ResultType): 출력 순서 수정

sangcci committed:
refactor($RestartTest): @ParameterizedTest name 속성 이용, 가독성 고려 DisplayName 수정

sangcci committed:
style($RestartTest): 안쓰는 import 삭제

sangcci committed:
docs($docs\README): 기능 목록 및 객체 설계안 수정

sangcci committed:
Merge pull request #3 from baksakcci/develop

도메인 구조 수정 및 테스트 코드 추가
